### DISCUSSION TITLE
Options to secure API

### DISCUSSION THREAD

**Discussion by lee-pai-long** [graphql-python/graphene-django#1479] [General]
**Is your feature request related to a problem? Please describe.**
I used [graphql-cop](https://github.com/dolevf/graphql-cop) to test my graphql API built using Graphene Django and the result is as follows:

```
[HIGH] Alias Overloading - Alias Overloading with 100+ aliases is allowed (Denial of Service - /graphql)
[HIGH] Directive Overloading - Multiple duplicated directives allowed in a query (Denial of Service - /graphql)
[HIGH] Field Duplication - Queries are allowed with 500 of the same repeated field (Denial of Service - /graphql)
[LOW] Field Suggestions - Field Suggestions are Enabled (Information Leakage - /graphql)
[MEDIUM] GET Method Query Support - GraphQL queries allowed using the GET method (Possible Cross Site Request Forgery (CSRF) - /graphql)
[HIGH] Introspection - Introspection Query Enabled (Information Leakage - /graphql)
[HIGH] Introspection-based Circular Query - Circular-query using Introspection (Denial of Service - /graphql)
[MEDIUM] POST based url-encoded query (possible CSRF) - GraphQL accepts non-JSON queries over POST (Possible Cross Site Request Forgery - /graphql)
```

I would like to have options for example to disable or limit use of aliases to prevent Alias Overloading but I can't find options to mitigate this or the other attacks.

**Describe the solution you'd like**
Is it possible to provide options to mitigate those attacks in a futur version of graphene-django ?

**Describe alternatives you've considered**
...

**Additional context**
...

**Comment by kiendang** [comment]
Is this something that could be solved with a custom [`ValidationRule`](https://docs.graphene-python.org/en/latest/execution/queryvalidation/)? You might want to take a look at #1475.

**Comment by lee-pai-long** [comment]
Hi @kiendang the instrospection can be solve with that yes, but not the rest.

**Comment by kiendang** [comment]
I think a few of these, Alias Overloading, Directive Overloading and Field Duplication could be solved by writing a custom `ValidationRule` that analyses the query.

The CSRF issues could be solved by configuring Django if I'm not wrong.

**Comment by lee-pai-long** [comment]
@kiendang but how to limit the number of aliases, directives and field in a custom validation rule ?

**Comment by kiendang** [comment]
I guess you could traverse the AST and count the number of occurrences? See [Implementing a custom `ValudationRule`](https://docs.graphene-python.org/en/latest/execution/queryvalidation/#implementing-custom-validators) and the [`Visitor` class](https://github.com/graphql-python/graphql-core/blob/v3.2.3/src/graphql/language/visitor.py).

Anw I'm converting this to a discussion which I think is more appropriate.

**Comment by lee-pai-long** [comment]
I created a feature request because I think security should be a first class citizen in graphene and graphene django. GraphQL is full of holes and I'm not sure I'm qualified to plug those holes correctly.

For example in JS Apollo server has a set of limits that one can easily configure: https://www.apollographql.com/docs/router/configuration/operation-limits/

**Reply by kiendang** [reply]
@lee-pai-long I agree that security enhancing features are legitimate requests. However this issue in its original state is under-spec-ed for a feature request. Examples of feature requests that are more actionable would be
- Provide a `ValidationRule` to limit the number of aliases.
- Provide a `ValidationRule` to disallow multiple duplicated directives.

There is nothing wrong with your original post, it just didn't yet specify specific things to implement in graphene-django. I just think a discussion would be a better place to spec things out before creating a feature request.

**Reply by lee-pai-long** [reply]
@kiendang thanks for your answer, ok I understand

**Reply by esabol** [reply]
So has _anyone_ written a custom `ValidationRule` to limit the number of aliases?

**Comment by kiendang** [comment]
@lee-pai-long thanks for providing references to [graphql-cop](https://github.com/dolevf/graphql-cop) and [Apollo](https://www.apollographql.com/docs/router/configuration/operation-limits/). Looks like it could be worth it to implement some of these as `ValidationRule`, or at least the alias limit one.

@sjdemartini you mentioned you are looking into validation for your GraphQL app? This could be of interest to you.

Also `ValidationRule` is part of [graphql-core](https://github.com/graphql-python/graphql-core), not Graphene. The [Strawberry framework](https://strawberry.rocks/) also uses graphql-core underneath and has an [extension](https://strawberry.rocks/docs/extensions/add-validation-rules) for specifying `ValidationRule` as well. It might be best to implement these rules in a separate repo/library without Graphene dependency so that it could be used by both frameworks. Pinging @erikwrede here. We could port the existing [`depth_limit_validator`](https://github.com/graphql-python/graphene/blob/master/graphene/validation/depth_limit.py) from Graphene to the new library as well.

**Reply by lee-pai-long** [reply]
HI @kiendang any idea of when this will be implemented ?

**Reply by kiendang** [reply]
no timeline at the moment

**Reply by lee-pai-long** [reply]
Hi @kiendang, any news about implementing this ?

**Reply by kiendang** [reply]
There's no one actively developing new features at the moment. The maintainers are currently only reviewing PRs, bug fixes.

**Comment by manjurulhoque** [comment]
Wrote an article on how anyone can use ValidationRule to implement
https://medium.com/@manzurulhoque/understanding-graphql-aliases-and-directives-a-security-perspective-c0049b913c8c

**Reply by esabol** [reply]
Thank you! Appreciate it.
